Python/ceval.c

```diff
@@ -5554,17 +5640,25 @@ make_coro(PyThreadState *tstate, PyFrameConstructor *con,
     return gen;
 }
 
+// If *steal_args* is set, the function will steal the references to all the arguments.
+// In case of error, the function returns null and if *steal_args* is set, the caller
+// will still own all the arguments.
 static InterpreterFrame *
 _PyEvalFramePushAndInit(PyThreadState *tstate, PyFrameConstructor *con,
                         PyObject *locals, PyObject* const* args,
-                        size_t argcount, PyObject *kwnames)
+                        size_t argcount, PyObject *kwnames, int steal_args)
 {
     InterpreterFrame * frame = _PyThreadState_PushFrame(tstate, con, locals);
     if (frame == NULL) {
         return NULL;
     }
     PyObject **localsarray = _PyFrame_GetLocalsArray(frame);
-    if (initialize_locals(tstate, con, localsarray, args, argcount, kwnames)) {
+    if (initialize_locals(tstate, con, localsarray, args, argcount, kwnames, steal_args)) {
+        if (steal_args) {
+            for (int i = 0; i < frame->stacktop; i++) {
+                Py_XINCREF(frame->localsplus[i]);
```

Comment:
Why is this necessary?

Reply:
So if the function fails the caller still owns all references. This simplifies a lot of failure scenarios. For instance, imagine that yo have a function that takes 5 parameters but you pass 10. In this case, the function will return NULL but it would have only stolen 5 parameters, while the caller is still in charge of decrementing the other 5. But the problem is that the caller needs to know if this was the problem or something else (because the error could have been something different). THis way, on failure, the caller always have to decrement the references, which is very easy to do in the normal paths

Reply:
When you say "the function" do you mean `_PyEvalFramePushAndInit` or `initialize_locals`?

Reply:
`_PyEvalFramePushAndInit ` with steal = 1

Reply:
But the argument can be extended to `initialize_locals` because `_PyEvalFramePushAndInit` doesn't know what went wrong if it returns `NULL`